{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b95ce8e7",
   "metadata": {},
   "source": [
    "#### Summary:\n",
    "This notebook contains code to prepare the inputs for Scrublet (MM files) and also reads in the output files and merges them into one to use later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0ecdb821",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(library(hdf5r))\n",
    "suppressMessages(library(Seurat))\n",
    "suppressMessages(library(Signac))\n",
    "suppressMessages(library(EnsDb.Hsapiens.v86))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressMessages(library(ggplot2))\n",
    "suppressMessages(library(Matrix))\n",
    "suppressMessages(library(harmony))\n",
    "suppressMessages(library(data.table))\n",
    "suppressMessages(library(ggpubr))\n",
    "suppressMessages(library(future))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c05e8208",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(library(tictoc))\n",
    "suppressMessages(library(plyr))\n",
    "suppressMessages(library(enrichR))\n",
    "suppressMessages(library(stringr))\n",
    "suppressMessages(library(ggrepel))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "aaaffffa",
   "metadata": {},
   "outputs": [],
   "source": [
    "### UPDATE THESE PATHS\n",
    "rds_indir = '/path/to/multiome/processing/pipeline/outputs/dir'\n",
    "outdir = \"/dir/to/save/doublets/list/to\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77ccaa73",
   "metadata": {},
   "source": [
    "# Exporting RDS Objects as MM files for Scrublet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7edbde28",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples <- c('R207','R217','R218','R221','R223',\n",
    "             'R226','R228','R234','R237','R238',\n",
    "             'R246', 'R247', 'R275', 'R277', 'R284', \n",
    "             'R290', 'R292', 'R316', 'R317', 'R319',\n",
    "             'R325','R326','R327','R332','R343',\n",
    "             'R349','R353','R354','R362','R363','R364')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8cf0ba3d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"R207 2022-08-25 11:59:34\"\n",
      "[1] \"R217 2022-08-25 12:02:55\"\n",
      "[1] \"R218 2022-08-25 12:06:10\"\n",
      "[1] \"R221 2022-08-25 12:08:01\"\n",
      "[1] \"R223 2022-08-25 12:09:46\"\n",
      "[1] \"R226 2022-08-25 12:13:02\"\n",
      "[1] \"R228 2022-08-25 12:16:00\"\n",
      "[1] \"R234 2022-08-25 12:18:36\"\n",
      "[1] \"R237 2022-08-25 12:21:33\"\n",
      "[1] \"R238 2022-08-25 12:24:23\"\n",
      "[1] \"R246 2022-08-25 12:27:13\"\n",
      "[1] \"R247 2022-08-25 12:28:09\"\n",
      "[1] \"R275 2022-08-25 12:30:32\"\n",
      "[1] \"R277 2022-08-25 12:32:15\"\n",
      "[1] \"R284 2022-08-25 12:34:51\"\n",
      "[1] \"R290 2022-08-25 12:37:05\"\n",
      "[1] \"R292 2022-08-25 12:39:55\"\n",
      "[1] \"R316 2022-08-25 12:42:40\"\n",
      "[1] \"R317 2022-08-25 12:46:24\"\n",
      "[1] \"R319 2022-08-25 12:50:43\"\n",
      "[1] \"R325 2022-08-25 12:54:43\"\n",
      "[1] \"R326 2022-08-25 12:55:32\"\n",
      "[1] \"R327 2022-08-25 12:59:36\"\n",
      "[1] \"R332 2022-08-25 13:01:56\"\n",
      "[1] \"R343 2022-08-25 13:04:40\"\n",
      "[1] \"R349 2022-08-25 13:06:15\"\n",
      "[1] \"R353 2022-08-25 13:07:39\"\n",
      "[1] \"R354 2022-08-25 13:09:55\"\n",
      "[1] \"R362 2022-08-25 13:12:06\"\n",
      "[1] \"R363 2022-08-25 13:15:12\"\n",
      "[1] \"R364 2022-08-25 13:18:28\"\n"
     ]
    }
   ],
   "source": [
    "for (sample in samples){\n",
    "    print(paste(sample, Sys.time()))\n",
    "    #set up necessary file structures -- this dir should be the sample's output dir from the multiome processing pipeline scripts\n",
    "    sample_dir <- file.path(rds_indir, sample)\n",
    "    mm_outdir <- file.path(sample_dir,'matrix_market')\n",
    "    dir.create(mm_outdir)\n",
    "    \n",
    "    #read in RDS\n",
    "    adata <- readRDS(file = file.path(sample_dir,\"final_filtered.rds\"))\n",
    "    \n",
    "    #pull out RNA counts to a separate object and writeMM\n",
    "    DefaultAssay(adata) <- 'RNA'\n",
    "    rna.counts <- GetAssayData(adata,slot='counts')\n",
    "    mm.fp <- file.path(mm_outdir,'matrix.mtx')\n",
    "    writeMM(rna.counts,mm.fp)\n",
    "    \n",
    "    #also export the gene list and barcodes lists\n",
    "    genes.fp <- file.path(mm_outdir,'genes.tsv')\n",
    "    write(row.names(rna.counts),genes.fp,sep='\\n')\n",
    "\n",
    "    barcodes.fp <- file.path(mm_outdir,'barcodes.tsv')\n",
    "    write(colnames(rna.counts),barcodes.fp,sep='\\n')\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86754c07",
   "metadata": {},
   "source": [
    "# Read in Scrublet outputs and merge into one file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a050227f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#samples for which I manually set a Scrublet threshold of 0.25\n",
    "redo_samples = c('R218','R221','R238','R275', 'R277',\n",
    "                 'R317', 'R319','R343','R349','R364')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "62a6919c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"R207\"             \"1662063138.78032\"\n",
      "[1] \"0.2557159026533267\"\n",
      "[1] \"R217\"             \"1662063138.81173\"\n",
      "[1] \"0.17592480680131617\"\n",
      "[1] \"R218\"             \"1662063138.86208\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R221\"             \"1662063138.88525\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R223\"             \"1662063138.91489\"\n",
      "[1] \"0.20057109796338418\"\n",
      "[1] \"R226\"             \"1662063138.95066\"\n",
      "[1] \"0.21798500551758931\"\n",
      "[1] \"R228\"             \"1662063138.98634\"\n",
      "[1] \"0.18229496603578663\"\n",
      "[1] \"R234\"             \"1662063139.05762\"\n",
      "[1] \"0.20616746937851066\"\n",
      "[1] \"R237\"             \"1662063139.10211\"\n",
      "[1] \"0.21591387236966297\"\n",
      "[1] \"R238\"             \"1662063139.15696\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R246\"             \"1662063139.19769\"\n",
      "[1] \"0.21450430034488194\"\n",
      "[1] \"R247\"             \"1662063139.24848\"\n",
      "[1] \"0.206563350297425\"\n",
      "[1] \"R275\"             \"1662063139.33675\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R277\"            \"1662063139.3739\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R284\"             \"1662063139.41574\"\n",
      "[1] \"0.17801217813100262\"\n",
      "[1] \"R290\"             \"1662063139.46352\"\n",
      "[1] \"0.17703093285732843\"\n",
      "[1] \"R292\"             \"1662063139.51499\"\n",
      "[1] \"0.24735820062933223\"\n",
      "[1] \"R316\"             \"1662063139.59036\"\n",
      "[1] \"0.17039126969717594\"\n",
      "[1] \"R317\"             \"1662063139.64933\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R319\"             \"1662063139.74982\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R325\"           \"1662063139.835\"\n",
      "[1] \"0.20025657670066127\"\n",
      "[1] \"R326\"             \"1662063139.96878\"\n",
      "[1] \"0.17792640580413635\"\n",
      "[1] \"R327\"             \"1662063140.03047\"\n",
      "[1] \"0.26261129679479317\"\n",
      "[1] \"R332\"             \"1662063140.10426\"\n",
      "[1] \"0.17897881757117817\"\n",
      "[1] \"R343\"             \"1662063140.20516\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R349\"             \"1662063140.25932\"\n",
      "[1] \"0.25\"\n",
      "[1] \"R353\"             \"1662063140.38611\"\n",
      "[1] \"0.19953090838268311\"\n",
      "[1] \"R354\"            \"1662063140.4369\"\n",
      "[1] \"0.2268391698711363\"\n",
      "[1] \"R362\"             \"1662063140.52182\"\n",
      "[1] \"0.27520463239107873\"\n",
      "[1] \"R363\"             \"1662063140.62373\"\n",
      "[1] \"0.16752320888746958\"\n",
      "[1] \"R364\"             \"1662063140.71958\"\n",
      "[1] \"0.25\"\n"
     ]
    }
   ],
   "source": [
    "fin_scrub_df = data.frame()\n",
    "\n",
    "for (sample in samples){\n",
    "    print(c(sample, Sys.time()))\n",
    "    #get the cutoff number\n",
    "    outdir = file.path(rds_indir, sample, 'matrix_market')\n",
    "    if (sample %in% redo_samples){\n",
    "        cutoff = '0.25'\n",
    "        print(cutoff)\n",
    "    } else {\n",
    "        files = list.files(outdir)\n",
    "        fp = files[grep('scrub',files)]\n",
    "        cutoff = substr(fp, unlist(gregexpr('cutoff',fp))+6, unlist(gregexpr('.txt',fp))-1)\n",
    "        print(cutoff)\n",
    "    }\n",
    "        \n",
    "    #read in scrublet file\n",
    "    scrub_fp = file.path(outdir,sprintf('scrublet_predicted_doublets_cutoff%s.txt',cutoff))\n",
    "    scrub_df = read.table(scrub_fp, sep='\\t', header=FALSE)\n",
    "    \n",
    "    #add on sample BC prefix\n",
    "    scrub_df$V1 = paste(sample, '_', scrub_df$V1, sep='')\n",
    "\n",
    "    #add to overall dataframe\n",
    "    fin_scrub_df = rbind(fin_scrub_df,scrub_df)\n",
    "}\n",
    "\n",
    "out_fp = file.path(outdir,'indiv_samples_combined.scrublet_predicted_doublets.txt')\n",
    "write.table(fin_scrub_df[,c(1,2,3)],out_fp, sep='\\t', row.names=FALSE, col.names=FALSE, quote=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "043d9b2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.2 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0\n",
       "LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] ggrepel_0.9.4             stringr_1.4.1            \n",
       " [3] enrichR_3.1               plyr_1.8.9               \n",
       " [5] tictoc_1.1                future_1.34.0            \n",
       " [7] ggpubr_0.5.0              data.table_1.14.8        \n",
       " [9] harmony_0.1.1             Rcpp_1.0.13              \n",
       "[11] Matrix_1.5-1              ggplot2_3.4.4            \n",
       "[13] dplyr_1.0.10              EnsDb.Hsapiens.v86_2.99.0\n",
       "[15] ensembldb_2.18.2          AnnotationFilter_1.18.0  \n",
       "[17] GenomicFeatures_1.46.1    AnnotationDbi_1.56.2     \n",
       "[19] Biobase_2.54.0            GenomicRanges_1.46.1     \n",
       "[21] GenomeInfoDb_1.30.1       IRanges_2.28.0           \n",
       "[23] S4Vectors_0.32.4          BiocGenerics_0.40.0      \n",
       "[25] Signac_1.5.0              sp_1.5-1                 \n",
       "[27] SeuratObject_4.1.2        Seurat_4.2.0             \n",
       "[29] hdf5r_1.3.7              \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] utf8_1.2.4                  reticulate_1.26            \n",
       "  [3] tidyselect_1.2.0            RSQLite_2.2.9              \n",
       "  [5] htmlwidgets_1.5.4           grid_4.1.1                 \n",
       "  [7] docopt_0.7.1                BiocParallel_1.28.3        \n",
       "  [9] Rtsne_0.16                  munsell_0.5.1              \n",
       " [11] codetools_0.2-18            ica_1.0-3                  \n",
       " [13] pbdZMQ_0.3-8                miniUI_0.1.1.1             \n",
       " [15] withr_3.0.1                 spatstat.random_3.0-1      \n",
       " [17] colorspace_2.1-1            progressr_0.11.0           \n",
       " [19] filelock_1.0.2              uuid_1.1-0                 \n",
       " [21] ROCR_1.0-11                 ggsignif_0.6.4             \n",
       " [23] tensor_1.5                  listenv_0.9.1              \n",
       " [25] MatrixGenerics_1.6.0        slam_0.1-50                \n",
       " [27] repr_1.1.4                  GenomeInfoDbData_1.2.7     \n",
       " [29] polyclip_1.10-4             bit64_4.0.5                \n",
       " [31] farver_2.1.2                parallelly_1.38.0          \n",
       " [33] vctrs_0.5.1                 generics_0.1.3             \n",
       " [35] BiocFileCache_2.2.0         lsa_0.73.3                 \n",
       " [37] ggseqlogo_0.1               R6_2.5.1                   \n",
       " [39] DelayedArray_0.20.0         bitops_1.0-8               \n",
       " [41] spatstat.utils_3.0-5        cachem_1.1.0               \n",
       " [43] assertthat_0.2.1            promises_1.2.0.1           \n",
       " [45] BiocIO_1.4.0                scales_1.2.1               \n",
       " [47] rgeos_0.5-9                 gtable_0.3.1               \n",
       " [49] globals_0.16.3              goftest_1.2-3              \n",
       " [51] rlang_1.1.4                 RcppRoll_0.3.0             \n",
       " [53] splines_4.1.1               rstatix_0.7.1              \n",
       " [55] rtracklayer_1.54.0          lazyeval_0.2.2             \n",
       " [57] broom_1.0.0                 spatstat.geom_2.4-0        \n",
       " [59] yaml_2.3.10                 reshape2_1.4.4             \n",
       " [61] abind_1.4-5                 backports_1.4.1            \n",
       " [63] httpuv_1.6.6                tools_4.1.1                \n",
       " [65] ellipsis_0.3.2              spatstat.core_2.4-4        \n",
       " [67] RColorBrewer_1.1-3          ggridges_0.5.4             \n",
       " [69] base64enc_0.1-3             progress_1.2.2             \n",
       " [71] zlibbioc_1.40.0             purrr_0.3.5                \n",
       " [73] RCurl_1.98-1.13             prettyunits_1.2.0          \n",
       " [75] rpart_4.1.19                deldir_2.0-4               \n",
       " [77] pbapply_1.6-0               cowplot_1.1.1              \n",
       " [79] zoo_1.8-12                  SummarizedExperiment_1.24.0\n",
       " [81] cluster_2.1.4               magrittr_2.0.3             \n",
       " [83] scattermore_0.8             lmtest_0.9-40              \n",
       " [85] RANN_2.6.1                  SnowballC_0.7.0            \n",
       " [87] ProtGenerics_1.26.0         fitdistrplus_1.1-8         \n",
       " [89] matrixStats_1.3.0           hms_1.1.2                  \n",
       " [91] patchwork_1.1.2             mime_0.12                  \n",
       " [93] evaluate_0.24.0             xtable_1.8-4               \n",
       " [95] XML_3.99-0.16               sparsesvd_0.2-1            \n",
       " [97] gridExtra_2.3               compiler_4.1.1             \n",
       " [99] biomaRt_2.50.3              tibble_3.1.8               \n",
       "[101] KernSmooth_2.23-20          crayon_1.5.3               \n",
       "[103] htmltools_0.5.8.1           mgcv_1.8-41                \n",
       "[105] later_1.3.2                 tidyr_1.2.1                \n",
       "[107] DBI_1.1.3                   tweenr_2.0.2               \n",
       "[109] dbplyr_2.2.1                MASS_7.3-58.1              \n",
       "[111] rappdirs_0.3.3              car_3.1-1                  \n",
       "[113] cli_3.6.3                   parallel_4.1.1             \n",
       "[115] igraph_1.3.5                pkgconfig_2.0.3            \n",
       "[117] GenomicAlignments_1.30.0    IRdisplay_1.1              \n",
       "[119] plotly_4.10.1               spatstat.sparse_3.1-0      \n",
       "[121] xml2_1.3.3                  XVector_0.34.0             \n",
       "[123] digest_0.6.36               sctransform_0.3.5          \n",
       "[125] RcppAnnoy_0.0.22            spatstat.data_3.1-2        \n",
       "[127] Biostrings_2.62.0           leiden_0.4.3               \n",
       "[129] fastmatch_1.1-3             uwot_0.1.14                \n",
       "[131] restfulr_0.0.15             curl_5.2.1                 \n",
       "[133] shiny_1.7.3                 Rsamtools_2.10.0           \n",
       "[135] rjson_0.2.21                lifecycle_1.0.4            \n",
       "[137] nlme_3.1-160                jsonlite_1.8.8             \n",
       "[139] carData_3.0-5               viridisLite_0.4.2          \n",
       "[141] fansi_1.0.6                 pillar_1.8.1               \n",
       "[143] lattice_0.20-45             KEGGREST_1.34.0            \n",
       "[145] fastmap_1.2.0               httr_1.4.4                 \n",
       "[147] survival_3.4-0              glue_1.7.0                 \n",
       "[149] qlcMatrix_0.9.7             png_0.1-8                  \n",
       "[151] bit_4.0.5                   ggforce_0.4.1              \n",
       "[153] stringi_1.8.2               blob_1.2.3                 \n",
       "[155] memoise_2.0.1               IRkernel_1.3.1             \n",
       "[157] irlba_2.3.5.1               future.apply_1.9.1         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.1",
   "language": "R",
   "name": "ir411"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
